modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/metadata/IgniteMdSelectivity.java

```diff
@@ -110,6 +121,646 @@ public Double getSelectivity(IgniteSortedIndexSpool rel, RelMetadataQuery mq, Re
+        ColumnStatistics colStat = getColStat(rel, mq, tbl, pred);
+
+        if (colStat == null)
+            return guessSelectivity(pred);
+
+        return estimateNullSelectivity(colStat);
+    }
+
+    /**
+     * Estimate selectivity for equals predicate.
+     *
+     * @param rel RElNode.
+     * @param mq RelMetadataQuery.
+     * @param tbl IgniteTable.
+     * @param pred RexNode with predicate.
+     *
+     * @return Selectivity.
+     */
+    private double estimateEqualsSelectivity(
+        RelNode rel,
+        RelMetadataQuery mq,
+        IgniteTable tbl,
+        RexNode pred) {
+        ColumnStatistics colStat = getColStat(rel, mq, tbl, pred);
+
+        if (colStat == null)
+            return guessSelectivity(pred);
+
+        RexLiteral val = getOperand(pred, RexLiteral.class);
+
+        if (val == null)
+            return guessSelectivity(pred);
+
+        BigDecimal comparableVal = toComparableValue(val);
+
+        if (comparableVal == null)
+            return guessSelectivity(pred);
+
+        return estimateEqualsSelectivity(colStat, comparableVal);
+    }
+
+    /**
+     * Compute selectivity for "is not null" condition.
+     *
+     * @param rel RelNode.
+     * @param mq  RelMetadataQuery.
+     * @param tbl IgniteTable.
+     * @param pred RexNode.
+     * @return Selectivity estimation.
+     */
+    private double estimateIsNotNullSelectivity(RelNode rel, RelMetadataQuery mq, IgniteTable tbl, RexNode pred) {
+        ColumnStatistics colStat = getColStat(rel, mq, tbl, pred);
+
+        if (colStat == null)
+            return guessSelectivity(pred);
+
+        return estimateNotNullSelectivity(colStat);
+    }
+
+    /**
+     * Estimate selectivity for comparison predicate.
+     *
+     * @param rel RelNode.
+     * @param mq RelMetadataQuery.
+     * @param tbl IgniteTable.
+     * @param pred RexNode.
+     * @return Selectivity.
+     */
+    private double estimateComparisonSelectivity(RelNode rel, RelMetadataQuery mq, IgniteTable tbl, RexNode pred) {
+        ColumnStatistics colStat = getColStat(rel, mq, tbl, pred);
+
+        if (colStat == null)
+            return guessSelectivity(pred);
+
+        return estimateRangeSelectivity(colStat, pred);
+    }
+
+    /**
+     * Get column statistics.
+     *
+     * @param rel RelNode.
+     * @param mq RelMetadataQuery.
+     * @param tbl IgniteTable to get statistics from.
+     * @param pred Predicate to get statistics by related column.
+     * @return ColumnStatistics or {@code null}.
+     */
+    private ColumnStatistics getColStat(RelNode rel, RelMetadataQuery mq, IgniteTable tbl, RexNode pred) {
+        SqlKind predKind = pred.getKind();
+
+        if (predKind != SqlKind.IS_NULL && predKind != SqlKind.IS_NOT_NULL && predKind != SqlKind.LOCAL_REF &&
+            predKind != SqlKind.NOT && !predKind.belongsTo(SqlKind.COMPARISON))
+            return null;
+
+        RexLocalRef operand = getOperand(pred, RexLocalRef.class);
```

Comment:
If we would handle `RexLocalRef` only, this whole machinery will be working for filters merged with table scans only. But beside those there are conditions from above of relational tree, e.g:
```
CorrelatedFilter
    Exchange
        TableScan
```